packages/application/src/index.ts

```diff
@@ -222,62 +246,133 @@ export class Application<T extends Widget> {
 
     // Resolve immediately if the plugin is already activated.
-    if (data.activated) {
-      return Promise.resolve(undefined);
+    if (plugin.activated) {
+      return;
     }
 
     // Return the pending resolver promise if it exists.
-    if (data.promise) {
-      return data.promise;
+    if (plugin.promise) {
+      return plugin.promise;
     }
 
     // Resolve the required services for the plugin.
-    let required = data.requires.map(t => this.resolveRequiredService(t));
+    const required = plugin.requires.map(t => this.resolveRequiredService(t));
 
     // Resolve the optional services for the plugin.
-    let optional = data.optional.map(t => this.resolveOptionalService(t));
-
-    // Create the array of promises to resolve.
-    let promises = required.concat(optional);
+    const optional = plugin.optional.map(t => this.resolveOptionalService(t));
 
     // Setup the resolver promise for the plugin.
-    data.promise = Promise.all(promises)
-      .then(services => {
-        return data.activate.apply(undefined, [this, ...services]);
-      })
+    plugin.promise = Promise.all([...required, ...optional])
+      .then(services => plugin!.activate.apply(undefined, [this, ...services]))
       .then(service => {
-        data.service = service;
-        data.activated = true;
-        data.promise = null;
+        plugin!.service = service;
+        plugin!.activated = true;
+        plugin!.promise = null;
       })
       .catch(error => {
-        data.promise = null;
+        plugin!.promise = null;
         throw error;
       });
 
     // Return the pending resolver promise.
-    return data.promise;
+    return plugin.promise;
+  }
+
+  /**
+   * Deactivate the plugin and its downstream dependents if and only if the
+   * plugin and its dependents all support `deactivate`.
+   *
+   * @param id - The ID of the plugin of interest.
+   *
+   * @returns A list of IDs of downstream plugins deactivated with this one.
+   */
+  async deactivatePlugin(id: string): Promise<string[]> {
+    // Reject the promise if the plugin is not registered.
+    const plugin = this._plugins.get(id);
+    if (!plugin) {
+      throw new ReferenceError(`Plugin '${id}' is not registered.`);
+    }
+
+    // Bail early if the plugin is not activated.
+    if (!plugin.activated) {
+      return [];
+    }
+
+    // Check that this plugin can deactivate.
+    if (!plugin.deactivate) {
+      throw new TypeError(`Plugin '${id}'#deactivate() method missing`);
+    }
+
+    // Find the optimal deactivation order for plugins downstream of this one.
+    const manifest = Private.findDependents(id, this._plugins, this._services);
+    const downstream = manifest.map(id => this._plugins.get(id)!);
+
+    // Check that all downstream plugins can deactivate.
+    for (const plugin of downstream) {
+      if (!plugin.deactivate) {
+        throw new TypeError(
+          `Plugin ${plugin.id}#deactivate() method missing (depends on ${id})`
+        );
```

Comment:
This exception doesn't seem to correspond to the PR description: "an optional deactivate method which plugins may decide to implement. If it is provided, it should undo all the changes the activate method made, including removing the added commands, menus, etc. to complement the current activate() method."

Reply:
Which part is problematic, do you mean because `deactivate` is optional?

This only happens if you try to deactivate something and either the thing you're deactivating or one of its downstream plugins cannot deactivate. Is this behavior incorrect?

Reply:
I guess there is an additional line further down in the OP that mentions it, so probably ok.